flent/runners.py

```diff
@@ -952,6 +961,132 @@ def find_binary(cls, host, interval, length, ip_version, local_bind=None,
         raise RuntimeError("No suitable Iperf binary found.")
 
 
+class ParseException(Exception):
+    pass
+
+
+class SsRunner(ProcessRunner):
+    """Runner for iterated `ss -t -i -p`. Depends on same partitial output
+    separationa and time stamping as TcRunner."""
+
+    ip_v4_addr_sub_re = "([0-9]{1,3}.){3}[0-9]{1,3}(:\d+)"
+    # ref.: to commented, untinkered version: ISBN 978-0-596-52068-7
+    ip_v6_addr_sub_re = "(?:(?:[A-F0-9]{1,4}:){7}[A-F0-9]{1,4}|(?=(?:[A-F0-9]{0,4}:){0,7}[A-F0-9]{0,4})(([0-9A-F]{1,4}:){1,7}|:)((:[0-9A-F]{1,4}){1,7}|:))(:\d+)"
+
+    time_re = re.compile(r"^Time: (?P<timestamp>\d+\.\d+)", re.MULTILINE)
+    cwnd_re = re.compile(r"cwnd:(?P<cwnd>\d+)", re.MULTILINE)
+    rtt_re = re.compile(r"rtt:(?P<rtt>\d+\.\d+/\d+\.\d+)", re.MULTILINE)
+    pid_re = re.compile(r"pid=(?P<pid>\d+)", re.MULTILINE)
+    ports_ipv4_re = re.compile(r"" + "(?P<src_p>" + ip_v4_addr_sub_re + ")" + "\s+" +
+                               "(?P<dst_p>" + ip_v4_addr_sub_re + ")")
+    ports_ipv6_re = re.compile(r"" + "(?P<src_p>" + ip_v6_addr_sub_re + ")" + "\s+" +
+                               "(?P<dst_p>" + ip_v6_addr_sub_re + ")", re.IGNORECASE)
+    ss_header_re = re.compile(r"" + "State\s+Recv-Q\s+Send-Q\s+Local")
+
+    src_p = []
+    dst_p = []
+    # ref.: upstream ss
+    ss_states = ["UNKNOWN", "ESTAB", "SYN-SENT", "SYN-RECV", "FIN-WAIT-1",
+                 "FIN-WAIT-2", "TIME-WAIT", "UNCONN", "CLOSE-WAIT", "LAST-ACK",
+                 "LISTEN", "CLOSING"]
+    ss_states_re_pat = "|".join(ss_states)
+
+    def __init__(self, *args, **kwargs):
+        super(SsRunner, self).__init__(*args, **kwargs)
+        self.ctrl_port = str(self._parent.ctrl_port)
+
+    def filter_np_parent(self, part):
+        sub_part = []
+        sub_parts = re.split(self.ss_states_re_pat, part)
+        sub_parts = [sp for sp in sub_parts if sp.strip() and not self.ss_header_re.search(sp)]
+
+        for sp in sub_parts:
+            pid = self.pid_re.search(part)
+            if None is pid:
+                continue
+            pid_str = pid.group('pid')
+
+            f_ports = self.ports_ipv4_re.search(sp)
+            if None is f_ports:
+                f_ports = self.ports_ipv6_re.search(sp)
+
+            if None is f_ports:
+                raise ParseException()
```

Comment:
Well, it also has the effect that if you have a valid subpart followed by an invalid one, you're going to discard the valid one (because you'll raise an exception on the second run through the loop). Not sure this is going to happen in practice, though, since you only get this far if the pid regex matches...